sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2466,20 +2489,26 @@ class Dataset[T] private[sql](
 
     field.dataType match {
       case IntegerType | LongType | DoubleType | FloatType | BooleanType | ByteType =>
-        val validity = allocator.buffer(numBytesOfBitmap(numOfRows))
+        val validityVector = new BitVector("validity", allocator)
+        val validityMutator = validityVector.getMutator()
+        validityVector.allocateNew(numOfRows)
+        validityMutator.setValueCount(numOfRows)
         val buf = allocator.buffer(numOfRows * field.dataType.defaultSize)
         var nullCount = 0
-        rows.foreach { row =>
+        rows.zipWithIndex.foreach { case (row, index) =>
```

Comment:
I'll leave this open for a day or so, so Xusen can have a look since he wrote the conversion code.  If you want to update go ahead or I can before I merge, no biggie.  I also would like to have `fillWithArrow` and `getAndSetToArrow` use the same data type cases to avoid duplication, but I can do that later.